### ISSUE TITLE
Okta Spring Boot starter doesn't work with GraalVM

### ISSUE THREAD

**Issue by mraible** [okta/okta-spring-boot#192]
Steps to reproduce:

Create a new Spring Boot app using [HTTPie](https://httpie.org):

```
http https://start.spring.io/starter.zip \
     dependencies==web,okta \
     packageName==com.okta.rest \
     name=spring-boot \
     type=maven-project \
     -o spring-boot.zip
```

Then, add a `HelloController`:

```java
package com.okta.rest.controller;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;

@RestController
public class HelloController {

    @GetMapping("/hello")
    public String hello(@AuthenticationPrincipal Principal principal) {
        return "Hello, " + principal.getName() + "!";
    }
}
```

Configure it to be an OAuth 2.0 resource server by adding an issuer to `application.properties`:

```
okta.oauth2.issuer=https://dev-133337.okta.com/oauth2/default
```

All the following should work at this point (I got an access token from oidcdebugger.com):

```
$ mvn spring-boot:run
$ http :8080/hello
$ TOKEN=eyJraWQiOiJxOE1QMjFNNHZCVmxOSkxGbFFWNlN...
$ http :8080/hello Authorization:"Bearer $TOKEN"
```

To add "build native image support", I followed [these docs](https://repo.spring.io/milestone/org/springframework/experimental/spring-graalvm-native-docs/0.8.0/spring-graalvm-native-docs-0.8.0.zip!/reference/index.html#_configure_the_maven_plugin).

I upgraded my app to use Spring Boot v2.4.0-M2.

```xml
<parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>
    <version>2.4.0-M2</version>
    <relativePath/> <!-- lookup parent from repository -->
</parent>
```

I updated my configuration to avoid proxies:

```java
@SpringBootApplication(proxyBeanMethods = false)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}
}
```

I added the milestone repos to my `pom.xml`:

```xml
<repositories>
    <repository>
        <id>spring-milestones</id>
        <name>Spring Milestones</name>
        <url>https://repo.spring.io/milestone</url>
    </repository>
</repositories>
<pluginRepositories>
    <pluginRepository>
        <id>spring-milestones</id>
        <name>Spring Milestones</name>
        <url>https://repo.spring.io/milestone</url>
    </pluginRepository>
</pluginRepositories>
```

I configured the Spring Boot Maven plugin:

```xml
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <configuration>
        <image>
            <builder>paketobuildpacks/builder:tiny</builder>
            <env>
                <BP_BOOT_NATIVE_IMAGE>1</BP_BOOT_NATIVE_IMAGE>
                <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                    -Dspring.native.remove-yaml-support=true
                    -Dspring.spel.ignore=true
                </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
            </env>
        </image>
    </configuration>
</plugin>
```

I added the Spring GraalVM dependency:

```xml
<dependency>
    <groupId>org.springframework.experimental</groupId>
    <artifactId>spring-graalvm-native</artifactId>
    <version>0.8.0</version>
</dependency>
```

And built my application.

```
./mvnw spring-boot:build-image
```

Then I tried to run it. 

```
docker run -p 8080:8080 docker.io/library/demo:0.0.1-SNAPSHOT
```

It fails with the following error:

```
2020-09-21 18:40:59.056  INFO 1 --- [           main] com.okta.rest.Application                : Starting Application using Java 11.0.8 on 089430f03b4c with PID 1 (/workspace/com.okta.rest.Application started by cnb in /workspace)
2020-09-21 18:40:59.056  INFO 1 --- [           main] com.okta.rest.Application                : No active profile set, falling back to default profiles: default
2020-09-21 18:40:59.098  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: @Configuration classes need to be marked as proxyBeanMethods=false. Found: [com.okta.spring.boot.oauth.OktaOAuth2ResourceServerAutoConfig]
2020-09-21 18:40:59.099  INFO 1 --- [           main] ConditionEvaluationReportLoggingListener : 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2020-09-21 18:40:59.099 ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanDefinitionStoreException: @Configuration classes need to be marked as proxyBeanMethods=false. Found: [com.okta.spring.boot.oauth.OktaOAuth2ResourceServerAutoConfig]
        at org.springframework.context.annotation.ConfigurationClassPostProcessor.enhanceConfigurationClasses(ConfigurationClassPostProcessor.java:436) ~[com.okta.rest.Application:na]
        at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanFactory(ConfigurationClassPostProcessor.java:273) ~[com.okta.rest.Application:na]
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:299) ~[na:na]
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:134) ~[na:na]
        at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:751) ~[na:na]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:569) ~[na:na]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144) ~[na:na]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:751) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:410) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:326) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1280) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1269) ~[com.okta.rest.Application:na]
        at com.okta.rest.Application.main(Application.java:10) ~[com.okta.rest.Application:na]                                                                    
```

**Comment by mraible** [comment]
I tried to switch to Spring Security's dependencies and configuration as a workaround. I replaced the Okta Spring Boot starter with:

```xml
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-config</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-resource-server</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-jose</artifactId>
</dependency>
```

And I changed `application.properties` to use Spring Security's key:

```properties
spring.security.oauth2.resourceserver.jwt.issuer-uri=https://dev-133320.okta.com/oauth2/default
```

I can run everything just fine with `mvn spring-boot:run`. It builds with GraalVM too. However, when I try to run it, I get:

```
2020-09-21 23:43:05.538  INFO 1 --- [           main] com.okta.rest.Application                : Starting Application using Java 11.0.8 on 94f7aaa2abf1 with PID 1 (/workspace/com.okta.rest.Application started by cnb in /workspace)
2020-09-21 23:43:05.539  INFO 1 --- [           main] com.okta.rest.Application                : No active profile set, falling back to default profiles: default
2020-09-21 23:43:05.566 ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed

java.lang.IllegalStateException: java.security.NoSuchAlgorithmException: RSA KeyFactory not available
        at org.springframework.security.converter.RsaKeyConverters.rsaFactory(RsaKeyConverters.java:129) ~[na:na]
        at org.springframework.security.converter.RsaKeyConverters.pkcs8(RsaKeyConverters.java:63) ~[na:na]
        at org.springframework.security.config.crypto.RsaKeyConversionServicePostProcessor.pkcs8(RsaKeyConversionServicePostProcessor.java:89) ~[na:na]
        at org.springframework.security.config.crypto.RsaKeyConversionServicePostProcessor.postProcessBeanFactory(RsaKeyConversionServicePostProcessor.java:66) ~[na:na]
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:299) ~[na:na]
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:185) ~[na:na]
        at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:751) ~[na:na]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:569) ~[na:na]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144) ~[na:na]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:751) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:410) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:326) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1280) ~[com.okta.rest.Application:na]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1269) ~[com.okta.rest.Application:na]
        at com.okta.rest.Application.main(Application.java:10) ~[com.okta.rest.Application:na]
Caused by: java.security.NoSuchAlgorithmException: RSA KeyFactory not available
        at java.security.KeyFactory.<init>(KeyFactory.java:138) ~[na:na]
        at java.security.KeyFactory.getInstance(KeyFactory.java:183) ~[na:na]
        at org.springframework.security.converter.RsaKeyConverters.rsaFactory(RsaKeyConverters.java:127) ~[na:na]
        ... 15 common frames omitted
```

Any ideas @jgrandja?

**Comment by mraible** [comment]
I tried using the `full` builder as [documented here](https://repo.spring.io/milestone/org/springframework/experimental/spring-graalvm-native-docs/0.8.0/spring-graalvm-native-docs-0.8.0.zip!/reference/index.html#_configure_the_maven_plugin). 

```xml
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <configuration>
        <image>
            <builder>paketobuildpacks/builder:full</builder>
            <env>
                <BP_BOOT_NATIVE_IMAGE>1</BP_BOOT_NATIVE_IMAGE>
                <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                    -Dspring.native.remove-yaml-support=true
                    -Dspring.spel.ignore=true
                </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
            </env>
        </image>
    </configuration>
</plugin>
```

Unfortunately, the same error still happens.

**Comment by mraible** [comment]
Hmmm, the [paketobuildpacks/builder](https://github.com/paketo-buildpacks/builder) repo says it's `full-cf` instead of `full`. Trying... (will update this comment with results so I don't keep spamming y'all).

**Comment by mraible** [comment]
It seems [I might need to pass in `--enable-https` as part of the build](https://www.graalvm.org/reference-manual/native-image/JCASecurityServices/). I tried this:

```xml
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <configuration>
        <image>
            <env>
                <BP_BOOT_NATIVE_IMAGE>1</BP_BOOT_NATIVE_IMAGE>
                <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                    -Dspring.native.remove-yaml-support=true
                    -Dspring.spel.ignore=true
                    --enable-https
                </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
            </env>
        </image>
    </configuration>
</plugin>
```

And it got me a bit further! Now I get:

```
2020-09-22 03:08:31.096  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration': Unsatisfied dependency expressed through method 'setFilterChainProxySecurityConfigurer' parameter 1; nested exception is org.springframework.beans.ConversionNotSupportedException: Failed to convert value of type 'java.lang.String' to required type 'java.util.List'; nested exception is java.lang.IllegalStateException: Cannot convert value of type 'java.lang.String' to required type 'org.springframework.security.config.annotation.SecurityConfigurer': no matching editors or conversion strategy found
```

I'll try downgrading.

**Comment by bdemers** [comment]
I'm not sure on the exact overlap between running graal native image directly or via the Spring plugin (i'm guessing they are similar).
But these are the options I needed to enable for the Okta CLI:
https://github.com/oktadeveloper/okta-cli/blob/master/cli/pom.xml#L205-L211

It's a much different use case though, and I remember pulling out my hair to get things working originally.
(NOTE: I'm only using a script (exec plugin) because the previous version of the Graal Native Image Maven Plugin didn't work correctly on Windows.

**Comment by mraible** [comment]
It looks like this will be fixed in Spring GraalVM Native 0.9.0, which is scheduled for December 18, 2020.

https://twitter.com/sdeleuze/status/1309491871769595904

**Comment by mraible** [comment]
From https://twitter.com/sdeleuze/status/1338508916171427841:

> We've just released Spring Native for GraalVM 0.8.4 based on SpringBoot 2.4.1. It introduces SpringSecurity, 
@RSocketIO, WebSocket and scheduling support. Enjoy!

**Comment by mraible** [comment]
I tried this again today with the Spring GraalVM Native [v0.8.4 release](https://github.com/spring-projects-experimental/spring-graalvm-native/releases/tag/0.8.4). Here's the steps I used.

Create a new Spring Boot app using [HTTPie](https://httpie.org):

```
http https://start.spring.io/starter.zip \
     dependencies==web,okta \
     packageName==com.okta.rest \
     name=spring-boot \
     type=maven-project \
     -o spring-boot.zip
```

Then, add a `HelloController`:

```java
package com.okta.rest.controller;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;

@RestController
public class HelloController {

    @GetMapping("/hello")
    public String hello(@AuthenticationPrincipal Principal principal) {
        return "Hello, " + principal.getName() + "!";
    }
}
```

Configure it to talk to Okta using the [Okta CLI](https://cli.okta.com):

```
okta apps create 
```

Select Web > Okta Spring Boot Starter.

I configured the Spring Boot Maven plugin:

```xml
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <configuration>
        <image>
            <builder>paketobuildpacks/builder:tiny</builder>
            <env>
                <BP_BOOT_NATIVE_IMAGE>1</BP_BOOT_NATIVE_IMAGE>
                <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                    --enable-all-security-services
                </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
            </env>
        </image>
    </configuration>
</plugin>
```

I added the Spring GraalVM dependency:

```xml
<dependency>
    <groupId>org.springframework.experimental</groupId>
    <artifactId>spring-graalvm-native</artifactId>
    <version>0.8.4</version>
</dependency>
```

And built my application.

```
./mvnw spring-boot:build-image
```

I got an error about it not being available in a repo, so I added milestone repos to my `pom.xml`:

```xml
<repositories>
    <repository>
        <id>spring-milestones</id>
        <name>Spring Milestones</name>
        <url>https://repo.spring.io/milestone</url>
    </repository>
</repositories>
<pluginRepositories>
    <pluginRepository>
        <id>spring-milestones</id>
        <name>Spring Milestones</name>
        <url>https://repo.spring.io/milestone</url>
    </pluginRepository>
</pluginRepositories>
```

Then I tried to build it again. It fails with a reasonable error.

```
[INFO]     [creator]     Excluding 103 auto-configurations from spring.factories file
[INFO]     [creator]     Processing spring.factories - EnableAutoConfiguration lists #6 configurations
[INFO]     [creator]     Fatal error:java.lang.IllegalStateException: java.lang.IllegalStateException: 
ERROR: in 'com.okta.spring.boot.oauth.OktaOAuth2AutoConfig' these methods are directly invoking methods marked 
@Bean: [oidcUserService] - due to the enforced proxyBeanMethods=false for components in a native-image, please 
consider refactoring to use instance injection. If you are confident this is not going to affect your 
application, you may turn this check off using -Dspring.native.verify=false.
```

Like before, I updated my configuration to avoid proxies:

```java
@SpringBootApplication(proxyBeanMethods = false)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}
}
```

Same error, so I tried `./mvnw spring-boot:build-image -Dspring.native.verify=false`. It still fails:

```
[INFO]     [creator]     Caused by: java.lang.IllegalStateException: 
ERROR: in 'com.okta.spring.boot.oauth.OktaOAuth2AutoConfig' these methods are directly invoking methods marked
@Bean: [oidcUserService] - due to the enforced proxyBeanMethods=false for components in a native-image, 
please consider refactoring to use instance injection. If you are confident this is not going to affect 
your application, you may turn this check off using -Dspring.native.verify=false.                                              
```

**Comment by mraible** [comment]
I tried switching to Spring Security's dependencies and configuration as a workaround:


```xml
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-config</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-resource-server</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-jose</artifactId>
</dependency>
```

Change `application.properties` to use Spring Security's key:


```properties
spring.security.oauth2.resourceserver.jwt.issuer-uri=https://dev-133320.okta.com/oauth2/default
```

Then, I tried:

```
./mvnw spring-boot:build-image
```

It worked! Next, let's see if it works:

```
docker run -p 8080:8080 docker.io/library/demo:0.0.1-SNAPSHOT
```

Nope! 

```
Caused by: java.net.MalformedURLException: Accessing an URL protocol that was not enabled. 
The URL protocol https is supported but not enabled by default. It must be enabled by adding the 
--enable-url-protocols=https option to the native-image command.
```

I updated my `pom.xml`:

```xml
<env>
    <BP_BOOT_NATIVE_IMAGE>1</BP_BOOT_NATIVE_IMAGE>
    <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
        --enable-all-security-services
        --enable-url-protocols=https
    </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
</env>
```

Now it builds and runs! 💥

I tested it and it *almost* works.

```
$ http :8080/hello # 401, generate token...
$ TOKEN=eyJraWQiOiJxOE1QMjFNNHZCVmxOSkxGbFFWNlN...
$ http :8080/hello Authorization:"Bearer $TOKEN"
```

This results in a 500:

```
HTTP/1.1 500
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Connection: close
Content-Type: application/json
Date: Tue, 15 Dec 2020 00:03:34 GMT
Expires: 0
Pragma: no-cache
Transfer-Encoding: chunked
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block

{
    "error": "Internal Server Error",
    "message": "",
    "path": "/hello",
    "status": 500,
    "timestamp": "2020-12-15T00:03:34.458+00:00"
}
```

In the Docker console, the error is:

```
SEVERE: Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.NullPointerException] with root cause
java.lang.NullPointerException
	at com.okta.rest.controller.HelloController.hello(HelloController.java:14)
	at java.lang.reflect.Method.invoke(Method.java:566)
```

**Comment by mraible** [comment]
Hmmm, the `NullPointerException` happens even when I run `mvn spring-boot:run`. 

@bdemers Any idea why `@AuthenticationPrincipal Principal principal` isn't resolving?

**Comment by mraible** [comment]
Changing my `HelloController` to the following fixed the NPE when running with `mvn spring-boot:run`:

```
package com.okta.rest.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;

@RestController
public class HelloController {

    @GetMapping("/hello")
    public String hello(Principal principal) {
        return "Hello, " + principal.getName() + "!";
    }
}
```

After I got that fixed, I re-built the image:

```
$ ./mvnw spring-boot:build-image
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:11 min
```

Ran it in Docker:

```
$ docker run -p 8080:8080 docker.io/library/demo:0.0.1-SNAPSHOT
...
2020-12-15 03:10:37.521  INFO 1 --- [           main] com.okta.rest.DemoApplication            
: Started DemoApplication in 0.978 seconds (JVM running for 0.982)
```

Then, in another terminal window:

```
$ http :8080/hello Authorization:"Bearer $TOKEN"
HTTP/1.1 200
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Connection: keep-alive
Content-Length: 28
Content-Type: text/plain;charset=UTF-8
Date: Tue, 15 Dec 2020 03:11:03 GMT
Expires: 0
Keep-Alive: timeout=60
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block

Hello, matt.raible@okta.com!
```

It works! 🎉

**Comment by mraible** [comment]
Next, I tried changing `@SpringBootApplication(proxyBeanMethods = false)` to `@SpringBootApplication`.

It built a few seconds faster: `04:07 min`. 

And it works, too! 💥

**Comment by sdeleuze** [comment]
Good news 👍🏼 For the error you see with Okta, I think you need to refactor `OktaOAuth2AutoConfig` to use `@Configuration(proxyBeanMethods = false)` (at library level in order to keep JVM compatibility and enable native support at the same time) and avoid using cross `@Bean` invocations as documented [here](https://repo.spring.io/milestone/org/springframework/experimental/spring-graalvm-native-docs/0.8.4/spring-graalvm-native-docs-0.8.4.zip!/reference/index.html#_limitations). Please let me know if you need some guidance.

**Comment by mraible** [comment]
@arvindkrishnakumar-okta Any updates on this? I'd love for it to be possible to build native images for Spring Boot apps that use our starter.

**Comment by mraible** [comment]
I tried this again today with Spring Boot 2.5.1, Spring Native [v0.10.0](https://github.com/spring-projects-experimental/spring-native/releases/tag/0.10.0), and the Okta Spring Boot starter v2.0.1. 

I don't expect a response from anyone on this, I just want to document the latest state of things. 🙂

Here's my pom.xml:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.5.1</version>
        <relativePath/> <!-- lookup parent from repository -->
    </parent>
    <groupId>com.example</groupId>
    <artifactId>demo</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>demo</name>
    <description>Demo project for Spring Boot</description>
    <properties>
        <java.version>11</java.version>
        <spring-native.version>0.10.0</spring-native.version>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>com.okta.spring</groupId>
            <artifactId>okta-spring-boot-starter</artifactId>
            <version>2.0.1</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.experimental</groupId>
            <artifactId>spring-native</artifactId>
            <version>${spring-native.version}</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <configuration>
                    <image>
                        <builder>paketobuildpacks/builder:tiny</builder>
                        <env>
                            <BP_NATIVE_IMAGE>true</BP_NATIVE_IMAGE>
                            <BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                                --enable-all-security-services
                                --enable-url-protocols=https
                            </BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS>
                        </env>
                    </image>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.springframework.experimental</groupId>
                <artifactId>spring-aot-maven-plugin</artifactId>
                <version>${spring-native.version}</version>
                <executions>
                    <execution>
                        <id>test-generate</id>
                        <goals>
                            <goal>test-generate</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>generate</id>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

    <repositories>
        <repository>
            <id>spring-releases</id>
            <name>Spring Releases</name>
            <url>https://repo.spring.io/release</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
    </repositories>
    <pluginRepositories>
        <pluginRepository>
            <id>spring-releases</id>
            <name>Spring Releases</name>
            <url>https://repo.spring.io/release</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </pluginRepository>
    </pluginRepositories>

</project>
```

I added the following to my `application.properties` file:

```properties
okta.oauth2.issuer=https://dev-1309757.okta.com/oauth2/default
```

When I build with `mvn spring-boot:build-image`, the error is:

```
[ERROR] Failed to execute goal org.springframework.experimental:spring-aot-maven-plugin:0.10.0:test-generate 
(test-generate) on project demo: Build failed during Spring AOT test code generation: ERROR: in 'com.okta.spring.boot.oauth.OktaOAuth2AutoConfig' these methods are directly invoking methods marked @
Bean: [oidcUserService] - due to the enforced proxyBeanMethods=false for components in a native-image, 
please consider refactoring to use instance injection. If you are confident this is not going to affect your application, 
you may turn this check off using -Dspring.native.verify=false. -> [Help 1]
```

Next, I replaced the Okta Spring Boot starter with Spring Security:

```xml
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-config</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-resource-server</artifactId>
</dependency>
<dependency>
    <groupId>org.springframework.security</groupId>
    <artifactId>spring-security-oauth2-jose</artifactId>
</dependency>
```

And changed my property name in `application.properties` to `spring.security.oauth2.resourceserver.jwt.issuer-uri`. 

I also added a resource server with Spring Security.

```java
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.annotation.web.configurers.oauth2.server.resource.OAuth2ResourceServerConfigurer;

@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http
            .authorizeRequests(request -> request.anyRequest().authenticated())
            .oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt);
    }
}
```

Then, I built my image:
```
./mvnw spring-boot:build-image
...
Execution time: 3 min. 51 s. 
```

And ran it with:
```
docker run -p 8080:8080 docker.io/library/demo:0.0.1-SNAPSHOT
```

Everything starts up:

```
2021-06-14 21:37:20.668  INFO 1 --- [           main] com.okta.rest.DemoApplication            : Started DemoApplication in 0.846 seconds (JVM running for 0.849)
```

And a call to `/hello` with a valid access token works as expected.

```
$ http :8080/hello Authorization:"Bearer $TOKEN"
HTTP/1.1 200
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Connection: keep-alive
Content-Length: 28
Content-Type: text/plain;charset=UTF-8
Date: Mon, 14 Jun 2021 21:38:37 GMT
Expires: 0
Keep-Alive: timeout=60
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block

Hello, matt.raible@okta.com!
```

**Comment by mraible** [comment]
This will be fixed by v2.1.1.
